activerecord/lib/active_record/connection_adapters/abstract/query_log_tags.rb

```diff
@@ -0,0 +1,205 @@
+        mattr_accessor :components, instance_accessor: false, default: [:application]
+        mattr_accessor :cache_query_log_tags, instance_accessor: false, default: true
+        mattr_accessor :backtrace_cleaner, default: ActiveSupport::BacktraceCleaner.new
+        thread_mattr_accessor :cached_comment, instance_accessor: false
+
+        class << self
+          # Updates the context used to construct the query log tags.
+          # Resets the cached comment if <tt>cache_query_log_tags</tt> is +true+.
+          def update(**options)
+            context.merge!(**options.symbolize_keys)
+            self.cached_comment = nil
+          end
+
+          def tags_available?
+            components.present? || inline_annotations.present?
+          end
+
+          # Returns an SQL comment +String+ containing the query log tags.
+          # Sets and returns a cached comment if <tt>cache_query_log_tags</tt> is +true+.
+          def comment
+            if cache_query_log_tags
+              self.cached_comment ||= uncached_comment
+            else
+              uncached_comment
+            end
+          end
+
+          def uncached_comment
+            content = tag_content
+            if content.present?
+              "/*#{escape_sql_comment(content)}*/"
+            else
+              ""
+            end
+          end
+
+          # Returns a +String+ containing any inline comments from +with_annotation+.
+          def inline_comment
+            return nil unless inline_annotations.present?
+            "/*#{escape_sql_comment(inline_tag_content)}*/"
+          end
+
+          # Manually clear the comment cache.
+          def clear_comment_cache!
+            self.cached_comment = nil
+          end
+
+          # Annotate any query within `&block`. Can be nested.
+          def with_annotation(comment, &block)
+            self.inline_annotations.push(comment)
+            block.call if block.present?
+          ensure
+            self.inline_annotations.pop
+          end
+
+          # Return the set of active inline annotations from +with_annotation+.
+          def inline_annotations
+            context[:inline_annotations] ||= []
+          end
+
+          # QueryLogTags +component+ methods
+
+          # Set during Rails boot in lib/active_record/railtie.rb
+          def application # :nodoc:
+            context[:application_name]
+          end
+
+          def pid # :nodoc:
+            Process.pid
+          end
+
+          def connection_config # :nodoc:
+            ActiveRecord::Base.connection_db_config
+          end
+
+          def socket # :nodoc:
+            connection_config.socket
+          end
+
+          def db_host # :nodoc:
+            connection_config.host
+          end
+
+          def database # :nodoc:
+            connection_config.database
+          end
+
+          def line # :nodoc:
+            backtrace_cleaner.add_silencer { |line| line.match?(/lib\/active_(record|support)/) }
```

Comment:
This line will add a new silencer to the backtrace cleaner every time the comment is built, which will cause it to both leak memory and get slower over time. Instead this should happen once, when the backtrace cleaner is created.

It looks like previously the application's backtrace cleaner was picked up and used here, but that was backed out because this line would mutate it - I think we should bring that back, given that we can't mutate the backtrace cleaner here anyway. In our app, we currently use `Marginalia::Comment.lines_to_ignore` to skip over some of our application directories (e.g. `lib/monkey_patches/`), and it would be great if we could retain that behaviour by configuring `Rails.backtrace_cleaner`.

It also seems like the `line` component won't work properly if `cache_query_log_tags` is enabled - I think this method should call `clear_comment_cache!`, overriding the caching even if it's configured.

Reply:
I've put back the application cleaner in the railtie and moved the silencer outside of the component method so it's fired once at startup.

I addressed the caching issue for `line` by disabling caching if `#line` is ever called (clearing the cache when it's called wouldn't help if the comment is already cached).

This may all change when I work on the configuration changes - proposal coming soon!

Reply:
Resolving as `line` has now been removed.